src/jrd/dfw.epp

```diff
@@ -1839,20 +1839,19 @@ static bool add_file(thread_db* tdbb, SSHORT phase, DeferredWork* work, jrd_tra*
 			// too low a value, make a stab at assigning one based on
 			// the indicated preference for the previous file length.
 
-			if ((start = X.RDB$FILE_START) < max)
+			start = MAX(min_start, X.RDB$FILE_START);
+
+			FOR(REQUEST_HANDLE handle2 TRANSACTION_HANDLE transaction)
+				FIRST 1 Y IN RDB$FILES
+					WITH Y.RDB$SHADOW_NUMBER EQ X.RDB$SHADOW_NUMBER
+					AND Y.RDB$FILE_SEQUENCE NOT MISSING
+					SORTED BY DESCENDING Y.RDB$FILE_SEQUENCE
 			{
-				FOR(REQUEST_HANDLE handle2 TRANSACTION_HANDLE transaction)
-					FIRST 1 Y IN RDB$FILES
-						WITH Y.RDB$SHADOW_NUMBER EQ X.RDB$SHADOW_NUMBER
-						AND Y.RDB$FILE_SEQUENCE NOT MISSING
-						SORTED BY DESCENDING Y.RDB$FILE_SEQUENCE
-				{
-					start = Y.RDB$FILE_START + Y.RDB$FILE_LENGTH;
-				}
-				END_FOR
+				min_start = Y.RDB$FILE_START + (Y.RDB$FILE_LENGTH ? Y.RDB$FILE_LENGTH : 1);
```

Comment:
I cannot comment on the fix, but is the test for `Y.RDB$FILE_LENGTH` about it being 0 or about it being `NULL`? If `NULL`, you should test `Y.RDB$FILE_LENGTH.NULL`...

Reply:
If the length of the previous file is not specified, `Y.RDB$FILE_LENGTH` is 0.

Reply:
It's nullable and it's better to check it explicitly. Right?

Reply:
Yes, `RDB$FILE_LENGTH` is a nullable field, but actually it can't be NULL here.

Reply:
In that case check for this will be very useful. IMHO this code is very far from performance critical paths, i.e. extra check does not cause troubles.